components/sup/src/protocols/types.rs

```diff
@@ -0,0 +1,243 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Generic types shared between the Supervisor and other components in the Habitat software
+//! ecosystem.
+//!
+//! Note: See `protocols/types.proto` for type level documentation for generated types.
+//! JW TODO: These types should be moved to the _core crate_ and where they will replace their
+//!          vanilla Rust type counterparts that we define there.
+
+use std::fmt;
+
+use hcore;
+use hcore::package::{self, Identifiable};
+
+pub use super::generated::types::*;
+use manager;
+
+impl fmt::Display for ApplicationEnvironment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}.{}", self.get_application(), self.get_environment())
+    }
+}
+
+impl fmt::Display for PackageIdent {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.has_version() && self.has_release() {
+            write!(
+                f,
+                "{}/{}/{}/{}",
+                self.get_origin(),
+                self.get_name(),
+                self.get_version(),
+                self.get_release()
+            )
+        } else if self.has_version() {
+            write!(
+                f,
+                "{}/{}/{}",
+                self.get_origin(),
+                self.get_name(),
+                self.get_version()
+            )
+        } else {
+            write!(f, "{}/{}", self.get_origin(), self.get_name())
+        }
+    }
+}
+
+impl fmt::Display for ProcessState {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let state = match *self {
+            ProcessState::Down => "down",
+            ProcessState::Up => "up",
+        };
+        write!(f, "{}", state)
+    }
+}
+
+impl fmt::Display for ServiceGroup {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut value = format!("{}.{}", self.get_service(), self.get_group());
+        if self.has_application_environment() {
+            value.insert_str(0, &format!("{}#", self.get_application_environment()));
+        }
+        if self.has_organization() {
+            value.push_str(&format!("@{}", self.get_organization()));
+        }
+        write!(f, "{}", value)
+    }
+}
+
+// JW TODO: These trait implementations to provide translations between protocol messages and
+// concrete Rust types defined in the core crate will go away eventually. We need to put the
+// core crate back into the Supervisor's repository and untangle our dependency hell before
+// that can happen.
+
```

Comment:
Are the implementations that are for types that _aren't_ defined in the `core` crate here to limit additional changes to the core of the Supervisor right now, and eventually we'll just use the protocol types directly, or is there another reason to have them?

(100% clear on the reason to do this for the `core` crate types, though.)

Reply:
It's an intermediary step. There's a few routes I'm looking at going, but the primary one is as follows:

* Move the generic socket stuff out into it's own crate. This will be something we can share between any network services and have that consuming network service define it's own dispatch table and codec.
* Improve the abstraction on SrvClient to be less generic and easier to call. Move this into `sup-ctl-client` (or something along those lines).
* Move the protocols out into `sup-protocols`
* Move Core back into this repository and have it depend upon `sup-protocols` replacing the basic types implementation with the protocol backed ones.
* Move the CLI handlers for `sup start`, `sup stop`, `sup unload`, `sup load` into the Hab CLI itself and that will depend upon `sup-ctl-client`.

Reply:
@reset Could you dump that list into an issue, so the tentative roadmap is more publicly-available?